libs/ui/src/virtual_keyboard/virtual_keyboard.tsx

```diff
@@ -206,29 +227,266 @@ export const SPACE_BAR_KEY: Key = {
+    const rightEdgeDistance = Math.abs(rightEdge - targetX);
+    return Math.min(leftEdgeDistance, rightEdgeDistance);
+  });
+
+  return closestEdges.indexOf(Math.min(...closestEdges));
+}
+
+/**
+ * Returns the index of the Key to focus if the user navigates to the previous row.
+ * Index of the Key is relative to its parent row only (not to all keys on the keyboard).
+ */
+function getPrevRowTargetButtonIndex(
+  keyMap: KeyMap,
+  focusedRowIndex: number,
+  rowRefs: React.MutableRefObject<Array<HTMLDivElement | null>>
+) {
+  return getAdjacentRowTargetButtonIndex(keyMap, rowRefs, focusedRowIndex, -1);
+}
+
+/**
+ * Returns the index of the Key to focus if the user navigates to the next row.
+ * Index of the Key is relative to its parent row only (not to all keys on the keyboard).
+ */
+function getNextRowTargetButtonIndex(
+  keyMap: KeyMap,
+  focusedRowIndex: number,
+  rowRefs: React.MutableRefObject<Array<HTMLDivElement | null>>
+) {
+  return getAdjacentRowTargetButtonIndex(keyMap, rowRefs, focusedRowIndex, 1);
+}
+
 export function VirtualKeyboard({
   onBackspace,
   onKeyPress,
+  onCancel,
+  onAccept,
   keyDisabled,
   keyMap = US_ENGLISH_KEYMAP,
+  enableWriteInAtiControllerNavigation,
 }: VirtualKeyboardProps): JSX.Element {
+  const [focusedRowIndex, setFocusedRowIndex] = useState(-1);
+
+  const keyMapWithActions: KeyMap = useMemo(() => {
+    const actions = enableWriteInAtiControllerNavigation
+      ? [
+          [SPACE_BAR_KEY, DELETE_KEY],
+          [CANCEL_KEY, ACCEPT_KEY],
+        ]
+      : // Cancel and Accept keys are rendered outside this component when ATI Controller navigation is off
+        [[SPACE_BAR_KEY, DELETE_KEY]];
+    return {
+      rows: [...keyMap.rows, ...actions],
+    };
+  }, [enableWriteInAtiControllerNavigation, keyMap.rows]);
+
+  const rowRefs = useRef<Array<HTMLDivElement | null>>([]);
+
+  // Remap the default behavior of the direction keys to navigate the keyboard grid in 2D
+  /* istanbul ignore next */
+  const handleKeyboardEventForVirtualKeyboard = useCallback(
+    (event: KeyboardEvent): void => {
+      switch (event.key) {
+        case Keybinding.PAGE_PREVIOUS:
+          advanceElementFocus(-1);
+          break;
+        case Keybinding.PAGE_NEXT:
+          advanceElementFocus(1);
+          break;
+        case Keybinding.FOCUS_PREVIOUS: {
+          const targetRowIndex = getPrevRowIndex(
+            keyMapWithActions,
+            focusedRowIndex
+          );
+          const targetKeyIndex = getPrevRowTargetButtonIndex(
+            keyMapWithActions,
+            focusedRowIndex,
+            rowRefs
+          );
+          const targetRow = rowRefs.current[targetRowIndex];
+          if (targetRow) {
+            const buttons = Array.from(targetRow.querySelectorAll('button'));
+            buttons[targetKeyIndex]?.focus();
+            setFocusedRowIndex(targetRowIndex);
+          }
+          preventBrowserScroll(event);
+          break;
+        }
```

Comment:
It might be a little more React-y to compute where to focus next from data rather than from the DOM. Since `keyMapWithActions` is a 2d array representing the keyboard, you could track the current focused element as a coordinate `(row, column)` in state and then pick out the next button to focus using those coordinates and the colspans defined in the key map. You could even attach a ref to each key in the key map.

I'm also fine with how it's currently implemented. It works and is clear enough. Just thought I'd throw out the idea

Reply:
I tried this initially but couldn't figure out a way to account for the gap between the keys defined [here](https://github.com/votingworks/vxsuite/pull/6071/files#diff-455a38a3cc5ca97747ac07039f626ce029a92b02f3f6f75f59e08af01e9deafeR32). Since we're mixing units (abstract column spans for width and pixels for gap) it seemed much easier to let the browser take care of rendering and then make the calculation after.

I agree the render-first approach isn't very React-y. The only way I could think to do the calculation before render would be to abandon the column span approach and give keys their own fixed pixel width, which has its own set of issues.

Reply:
Ah yeah I see what you mean. I just looked more closely at the keyboard and realized that the keys don't line up easily into columns. I think given that design, I see why we're constrained to this solution. I do think there's a case to make that the keys could be instead visually grouped into columns for clarity in navigation (since the normal keyboard layout is optimized for fingers, not controllers), but given that it's already finished I don't think it's worth changing.